File: majority_element/MajorityElement.java
import java.util.HashMap;
import java.util.Map;
import java.util.Arrays;

class MajorityElement {

    public static void main(String[] args) {
        int[] numbers = new int[] { 2, 2, 1, 1, 1, 2, 2 };
        int result = 2;

        System.out.printf("numbers: %s, result: %d\n", Arrays.toString(numbers), result);
        System.out.printf("Hash表: %d\n", hashTraversing(numbers));
        System.out.printf("数组暴力解法: %d\n", arrayViolence(numbers));
        System.out.printf("排序法: %d\n", sortMethod(numbers));
        System.out.printf("摩尔投票法: %d\n", boyerMoore(numbers));
    }

    private static int hashTraversing(int[] nums) {
        Map<Integer, Integer> hashMap = new HashMap<>();
        int majority = nums.length / 2;
        for (Integer v : nums) {
            Integer count = hashMap.get(v);
            count = (count == null ? 1 : ++count);
            hashMap.put(v, count);

            if (count > majority) {
                return v;
            }
        }
        throw new IllegalArgumentException("Can't find it!");
    }

    private static int arrayViolence(int[] nums) {
        int majority = nums.length / 2;
        for (int v : nums) {
            int count = 0;
            for (int num : nums) {
                if (v == num) {
                    count++;
                }
                if (count > majority) {
                    return v;
                }
            }
        }
        throw new IllegalArgumentException("Can't find it!");
    }

    private static int sortMethod(int[] nums) {
        Arrays.sort(nums);
        return nums[nums.length / 2];
    }

    private static int boyerMoore(int[] nums) {
        int count = 0;
        Integer currNum = null;
        for (int v : nums) {
            if (count == 0) {
                currNum = v;
            }
            count += currNum == v ? 1 : -1;
        }
        return currNum;
    }

}

File: majority_element/majority_element.go
package main

import (
	"fmt"
	"sort"
)

func main() {
	numbers := []int{2, 2, 1, 1, 1, 2, 2}
	result := 2

	fmt.Printf("numbers: %v, result: %d\n", numbers, result)
	fmt.Printf("Hash表: %d\n", oneHashTraversing(numbers))
	fmt.Printf("Hash表，两个for循环: %d\n", twoHashTraversing(numbers))
	fmt.Printf("数组暴力破解: %d\n", arrayViolence(numbers))
	fmt.Printf("排序法: %d\n", sortMethod(numbers))
	fmt.Printf("摩尔投票法: %d\n", boyerMoore(numbers))
}

func oneHashTraversing(nums []int) int {
	hash_table := make(map[int]int)
	majority := len(nums) / 2
	for _, v := range nums {
		hash_table[v]++
		if hash_table[v] > majority {
			return v
		}
	}

	panic("Can't find it!")
}

func twoHashTraversing(nums []int) int {
	hash_table := make(map[int]int)
	majority := len(nums) / 2
	for _, v := range nums {
		hash_table[v]++
	}

	for k, v := range hash_table {
		if v > majority {
			return k
		}
	}

	panic("Can't find it!")
}

func arrayViolence(nums []int) int {
	len := len(nums)
	majority := len / 2
	for i := 0; i < len; i++ {
		count := 0
		for j := 0; j < len; j++ {
			if nums[i] == nums[j] {
				count++
			}

			if count > majority {
				return nums[i]
			}
		}
	}

	panic("Can't find it!")
}

func sortMethod(nums []int) int {
	// 不推荐使用，既然是算法，还是自己写吧！
	sort.Ints(nums)
	return nums[len(nums) / 2]
}

func boyerMoore(nums []int) int {
	count := 0
	var curr_num int
	for _, v := range nums {
		if count == 0 {
			curr_num = v
		}
		if v == curr_num {
			count++
		} else {
			count--
		}
	}
	return curr_num
}